browser/tor/tor_control.cc

```diff
@@ -0,0 +1,1512 @@
+    owner_task_runner_->PostTask(FROM_HERE, base::BindOnce(std::move(callback),
+                                                           true, ""));
+    return;
+  }
+  owner_task_runner_->PostTask(FROM_HERE, base::BindOnce(std::move(callback),
+                                                         false, *version));
+}
+
+void TorControl::GetSOCKSListeners(
+    base::OnceCallback<void(
+        bool error, const std::vector<std::string>& listeners)> callback) {
+  std::unique_ptr<std::vector<std::string>> listeners =
+    std::make_unique<std::vector<std::string>>();
+  std::vector<std::string>* listeners_p = listeners.get();
+  Cmd("GETINFO net/listeners/socks",
+      base::BindRepeating(&TorControl::GetSOCKSListenersLine,
+                          base::Unretained(this), listeners_p),
+      base::BindOnce(&TorControl::GetSOCKSListenersDone,
+                     base::Unretained(this),
+                     std::move(listeners), std::move(callback)));
+}
+
+void TorControl::GetSOCKSListenersLine(std::vector<std::string>* listeners,
+                                           const std::string& status,
+                                           const std::string& reply) {
+  const char prefix[] = "net/listeners/socks=";
+  if (status != "250" ||
+      !base::StartsWith(reply, prefix, base::CompareCase::SENSITIVE)) {
+    VLOG(0) << "tor: unexpected `GETINFO net/listeners/socks' reply";
+    return;
+  }
+  listeners->push_back(reply.substr(strlen(prefix)));
+}
+
+void TorControl::GetSOCKSListenersDone(
+    std::unique_ptr<std::vector<std::string>> listeners,
+    base::OnceCallback<void(
+        bool error,
+        const std::vector<std::string>& listeners)> callback,
+    bool error, const std::string& status, const std::string& reply) {
+  if (error ||
+      status != "250" ||
+      reply != "OK" ||
+      listeners->empty()) {
+    owner_task_runner_->PostTask(FROM_HERE,
+                                 base::BindOnce(std::move(callback),
+                                                true,
+                                                std::vector<std::string>()));
+    return;
+  }
+  owner_task_runner_->PostTask(FROM_HERE, base::BindOnce(std::move(callback),
+                                                         false, *listeners));
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// Writing state machine
+
+// StartWrite()
+//
+//      Pick a write off the queue and start an I/O buffer for it.
+//
+//      Caller must ensure writing_ is true.
+//
+void TorControl::StartWrite() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(io_sequence_checker_);
+  DCHECK(writing_);
+  DCHECK(!writeq_.empty());
+  DCHECK(!cmdq_.empty());
+  auto buf = base::MakeRefCounted<net::StringIOBuffer>(writeq_.front());
+  writeiobuf_ = base::MakeRefCounted<net::DrainableIOBuffer>(buf, buf->size());
+  writeq_.pop();
+}
+
+// XXX DEBUG
+static std::string escapify(const char *buf, int len) {
```

Comment:
this should be at the top in an anonymous namespace

Reply:
fixed in 12c003763561ef5dc1ef014d4769d24a84c58883